concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+Once the program is compiled(/analyzed), there's no way to add new members to any of these types,
+and no reason to consider the idea of updating any of the cardinality accounting.
+
+What if we want "open" membership, where I *don't* know all of the possible
+types of values that will be part of a type in advance?
+
+This "open membership" is commonly known as **interfaces**.
+
+### interfaces
+
+Interfaces have the concept of "open membership", and are also typically
+combined with some definition of "features" that are required if something
+will be considered to implement the interface.
+
+In programming languages, the definition of "features" usually is expressed in
+terms of a set of methods.  This makes sense for programming languages,
+because if you don't know what methods you can call on a thing,
+you can't really do much with the thing!
+
+One could also imagine using a different definition of "features" to describe
+"interfaces" that apply purely to data:
+For example, stating that some map has the features to conform to an interface
+only if some entries of some known names and value types are present in the map.
+(However, this is not often seen, so this document will continue to frequently
+refer to "sets of methods" when discussing interfaces.)
+
+### explicit versus structural interfaces
+
+Interfaces as found in programming languages tend to fall into one of two large camps:
+*explicit* interfaces, or *structural* interfaces.
+
+Explicit interfaces require types implementing them to say so.
+
+Structural interfaces state some set of methods (or more generally, features)
+that must be present on a type for it to satisfy the interface;
+then, any type that has those methods implements the interface.
+
+Both explicit and structural styles of interfaces have "open" membership --
+you can "compile" a program that has interfaces in it,
+then add more libraries when you "link" the program and execute it,
+and those libraries can add more implementors of the interfaces!
+
+The explicit and structural styles vary in other practical details.
+For example, explicit interfaces require referencing the interface they implement;
+this makes them significantly less flexible, and creates many dependency management requirements.
+By contrast, structural interfaces can knowingly conform to an interfaces contract
+while still not referencing it directly; this can be extremely useful in
+designing systems for independently upgradeable pieces.
+
+(There are other yet more detailed ways in which these styles vary, but
+since they're not really important in this document, we'll rest here.)
+
+
+Types in the Wild
+-----------------
+
+These different categories of types are present in most programming languages,
+but not all of them are present in every language (!),
+and they often go by different names in different programming communities.
+Here's a few (non-exhaustive) examples in a handful of languages:
+
+### enums in the wild
```

Comment:
How exhaustive do you want to be with languages here?

JavaScript doesn't strictly have enums but can be [emulated in a variety of ways](https://2ality.com/2020/01/enum-pattern.html), TypeScript has them.

Reply:
Not even remotely exhaustive.  :)

(Instanceof https://github.com/ipld/specs/pull/242#issuecomment-600560619 .)

JS and TypeScript might be nice to comment on because they're pretty large-audience.  (And TypeScript, IFAICT, actually maps *really* closely to a bunch of these concepts... which makes it doubly-fascinating, since javascript so very much doesn't, and yet TypeScript is positioned as being so close to javascript!  mind-:boom:!)  But they're a bit tricky to talk about, so if we want to try, it probably deserves its own PR, at minimum.

(That link to how to emulate them is a great piece of content to include if we do expand this, though.)